caluma/caluma_form/signals.py

```diff
@@ -19,3 +31,154 @@ def set_historical_answer_type(sender, **kwargs):
         # revision.
         previous = historical_answer.get_previous_by_history_date()
         historical_answer.history_question_type = previous.history_question_type
+
+
+@receiver(post_delete, sender=models.Question)
+def cleanup_default_answer(sender, instance, **kwargs):
+    """Ensure default_answers are cleanedup."""
+    if instance.default_answer is not None:
+        instance.default_answer.delete()
+
+
+@receiver(post_init, sender=models.Document)
+def set_document_family(sender, instance, **kwargs):
+    """
+    Ensure document has the family pointer set.
+
+    This sets the document's family if not overruled or set already.
+    The family will be in the corresponding mutations when a tree
+    structure is created or restructured.
+    """
+    if instance.family_id is None:
+        # can't use instance.set_family() here as the instance isn't
+        # in DB yet, causing integrity errors
+        instance.family = instance
+
+
+def _update_calc_dependents(slug, old_expr, new_expr):
+    jexl = QuestionJexl()
+    old_q = set(jexl.extract_referenced_questions(old_expr))
+    new_q = set(jexl.extract_referenced_questions(new_expr))
+
+    to_add = new_q - old_q
+    to_remove = old_q - new_q
+
+    questions = models.Question.objects.filter(pk__in=list(to_add | to_remove))
+
+    for question in questions:
+        deps = set(question.calc_dependents)
+        if question.slug in to_add:
+            deps.add(slug)
+        else:
+            deps.remove(slug)
+        question.calc_dependents = list(deps)
+        question.save()
+
+
+def _update_or_create_calc_answer(question, document):
+    root_doc = document.family
+    root_form = document.family.form
+
+    struc = structure.FieldSet(root_doc, root_form)
+
+    for field in struc.get_fields(question.slug):
+        jexl = QuestionJexl(
+            {
+                "form": field.form,
+                "document": field.document,
+                "structure": field.parent(),
+            }
+        )
+
+        value = jexl.evaluate(field.question.calc_expression, raise_on_error=False)
```

Comment:
Why is it safe to ignore errors here? Might deserve a comment.

Reply:
Fair point, will do that.

Reply:
@sbor23 this is still missing, right?

Reply:
Ah true this went missing during rebase